### ISSUE TITLE
Build Pytorch Android from Source Failed with MSB4184 Error during Cmake

### ISSUE THREAD

**Issue by maxmx911** [pytorch/pytorch#39937]
I'm trying to build Putorch Android from Source and was following the guide from [here](https://pytorch.org/mobile/android/#building-pytorch-android-from-source) and I'm getting this error:
```
CMake Error at CMakeLists.txt:24 (project):
  Failed to run MSBuild command:

    C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/MSBuild/15.0/Bin/MSBuild.exe
  
  to get the value of VCTargetsPath:

    Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
    Copyright (C) Microsoft Corporation. All rights reserved.

    Build started 6/12/2020 4:42:05 PM.
    Project "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" on node 1 (default targets).
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Platforms\Tegra-Android\PlatformToolsets\Toolchain.props(18,5): error MSB4184: The expression """.Substring(0, -1)" cannot be evaluated. Length cannot be less than zero. Parameter name: length [D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj]
    Done Building Project "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" (default targets) -- FAILED.

    Build FAILED.

    "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" (default target) (1) ->
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Platforms\Tegra-Android\PlatformToolsets\Toolchain.props(18,5): error MSB4184: The expression """.Substring(0, -1)" cannot be evaluated. Length cannot be less than zero. Parameter name: length [D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj]

        0 Warning(s)
        1 Error(s)

    Time Elapsed 00:00:00.08

  Exit code: 1

-- Configuring incomplete, errors occurred!
```
I'm not sure what is being left out that leads to `"".Substring(0, -1)` cannot be evaluated. 

I'm using Git Bash on Windows 10 to execute the following:
```
export GRADLE_HOME=/c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1
export ANDROID_NDK=/c/Users/Sean_Ng/AppData/Local/Android/Sdk/ndk/21.1.6352462
export ANDROID_HOME=/c/Users/Sean_Ng/AppData/Local/Android/Sdk
sh ./scripts/build_pytorch_android.sh
```

This is the full output message: 
```
$ sh ./scripts/build_pytorch_android.sh
+++ dirname ./scripts/build_pytorch_android.sh
++ cd ./scripts/..
++ pwd -P
+ PYTORCH_DIR=/d/Coding/Android_Codes/pytorch
+ PYTORCH_ANDROID_DIR=/d/Coding/Android_Codes/pytorch/android
+ WORK_DIR=/d/Coding/Android_Codes/pytorch
+ echo PYTORCH_DIR:/d/Coding/Android_Codes/pytorch
PYTORCH_DIR:/d/Coding/Android_Codes/pytorch
+ echo WORK_DIR:/d/Coding/Android_Codes/pytorch
WORK_DIR:/d/Coding/Android_Codes/pytorch
+ echo ANDROID_HOME:/c/Users/Sean_Ng/AppData/Local/Android/Sdk
ANDROID_HOME:/c/Users/Sean_Ng/AppData/Local/Android/Sdk
+ '[' -z /c/Users/Sean_Ng/AppData/Local/Android/Sdk ']'
+ '[' '!' -d /c/Users/Sean_Ng/AppData/Local/Android/Sdk ']'
+ GRADLE_PATH=gradle
+ GRADLE_NOT_FOUND_MSG='Unable to find gradle, please add it to PATH or set GRADLE_HOME'
++ command -v gradle
+ '[' '!' -x '' ']'
+ '[' -z /c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1 ']'
+ GRADLE_PATH=/c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1/bin/gradle
+ '[' '!' -f /c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1/bin/gradle ']'
+ echo GRADLE_PATH:/c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1/bin/gradle
GRADLE_PATH:/c/Users/Sean_Ng/.gradle/wrapper/dists/gradle-5.4.1-all/3221gyojl5jsh0helicew7rwx/gradle-5.4.1/bin/gradle
+ ABIS_LIST=armeabi-v7a,arm64-v8a,x86,x86_64
+ CUSTOM_ABIS_LIST=false
+ '[' 0 -gt 0 ']'
+ echo ABIS_LIST:armeabi-v7a,arm64-v8a,x86,x86_64
ABIS_LIST:armeabi-v7a,arm64-v8a,x86,x86_64
+ LIB_DIR=/d/Coding/Android_Codes/pytorch/android/pytorch_android/src/main/jniLibs
+ INCLUDE_DIR=/d/Coding/Android_Codes/pytorch/android/pytorch_android/src/main/cpp/libtorch_include
+ mkdir -p /d/Coding/Android_Codes/pytorch/android/pytorch_android/src/main/jniLibs
+ mkdir -p /d/Coding/Android_Codes/pytorch/android/pytorch_android/src/main/cpp/libtorch_include
++ echo armeabi-v7a,arm64-v8a,x86,x86_64
++ tr , '\n'
+ for abi in $(echo $ABIS_LIST | tr ',' '\n')
+ echo abi:armeabi-v7a
abi:armeabi-v7a
+ OUT_DIR=/d/Coding/Android_Codes/pytorch/build_android_armeabi-v7a
+ rm -rf /d/Coding/Android_Codes/pytorch/build_android_armeabi-v7a
+ mkdir -p /d/Coding/Android_Codes/pytorch/build_android_armeabi-v7a
+ pushd /d/Coding/Android_Codes/pytorch
/d/Coding/Android_Codes/pytorch /d/Coding/Android_Codes/pytorch
+ python /d/Coding/Android_Codes/pytorch/setup.py clean
Building wheel torch-1.6.0a0+c3d4053
running clean
+ ANDROID_ABI=armeabi-v7a
+ /d/Coding/Android_Codes/pytorch/scripts/build_android.sh
Build with ANDROID_ABI[armeabi-v7a], ANDROID_NATIVE_API_LEVEL[21]
Bash: GNU bash, version 4.4.23(1)-release (x86_64-pc-msys)
Caffe2 path: /d/Coding/Android_Codes/pytorch
Using Android NDK at /c/Users/Sean_Ng/AppData/Local/Android/Sdk/ndk/21.1.6352462
Android NDK version: 21
-- Building for: Visual Studio 15 2017
CMake Error at CMakeLists.txt:24 (project):
  Failed to run MSBuild command:

    C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/MSBuild/15.0/Bin/MSBuild.exe

  to get the value of VCTargetsPath:

    Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
    Copyright (C) Microsoft Corporation. All rights reserved.

    Build started 6/12/2020 4:42:05 PM.
    Project "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" on node 1 (default targets).
    C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Platforms\Tegra-Android\PlatformToolsets\Toolchain.props(18,5): error MSB4184: The expression """.Substring(0, -1)" cannot be evaluated. Length cannot be less than zero. Parameter name: length [D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj]
    Done Building Project "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" (default targets) -- FAILED.

    Build FAILED.

    "D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj" (default target) (1) ->
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Platforms\Tegra-Android\PlatformToolsets\Toolchain.props(18,5): error MSB4184: The expression """.Substring(0, -1)" cannot be evaluated. Length cannot be less than zero. Parameter name: length [D:\Coding\Android_Codes\pytorch\build_android\CMakeFiles\3.15.2\VCTargetsPath.vcxproj]

        0 Warning(s)
        1 Error(s)

    Time Elapsed 00:00:00.08

  Exit code: 1

-- Configuring incomplete, errors occurred!
See also "D:/Coding/Android_Codes/pytorch/build_android/CMakeFiles/CMakeOutput.log".
```

I'd appreciate it if anyone can provide me some insights on this. Thank you!

cc @malfet @peterjc123 @nbcsm @guyang3532

**Comment by peterjc123** [comment]
For Android builds, you should use GCC or Clang instead of MSVC. Since I'm not so familiar with the process myself, I could not give much advice.

**Comment by maxmx911** [comment]
I printed out the `"${CMAKE_ARGS[@]}"` in `build_android.sh` and tried passing in the arguments one by one. 

`-DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake` 
This argument is causing the `The expression """.Substring(0, -1)" cannot be evaluated.` error. I think is the **.** in the path name. 

`-DANDROID_CPP_FEATURES=rtti exceptions`
The space in this argument would also cause error. 

Not passing them in I would encounter error later tho. Anyway, I installed Ubuntu on my VM and it builds successfully, so i guess Windows or the MSVC is not playing well with it. Thank you for your time on this tho! I'll close this issue.
